test/integration/short/client-pool-tests.js

```diff
@@ -641,6 +658,68 @@ describe('Client', function () {
     });
     it('should handle distance changing load balancing policies', changingDistancesTest('2'));
     it('should handle distance changing load balancing policies for control connection host', changingDistancesTest('1'));
+    [
+      new policies.speculativeExecution.NoSpeculativeExecutionPolicy(),
+      new ConstantSpeculativeExecutionPolicy(100, 1)
+    ].forEach(function (policy) {
+      context('with ' + policy.constructor.name, function () {
+        afterEach(function (done) {
+          helper.ccmHelper.resumeNode(1, done);
+        });
+        afterEach(function (done) {
+          helper.ccmHelper.resumeNode(2, done);
+        });
+        it('should wait until is completed on the first node', function (done) {
+          var client = newInstance({
+            pooling: { warmup: true },
+            policies: {
+              speculativeExecution: policy,
+              loadBalancing: new OrderedLoadBalancingPolicy(),
+              retry: new helper.FallthroughRetryPolicy()
+            },
+            socketOptions: {
+              readTimeout: 1000
+            }
+          });
+          utils.series([
+            client.connect.bind(client),
+            helper.toTask(helper.ccmHelper.pauseNode, null, 1),
+            function query(next) {
+              if (!(policy instanceof ConstantSpeculativeExecutionPolicy)) {
+                // Resume first node after a few ms
+                setTimeout(function () {
+                  helper.ccmHelper.resumeNode(1);
+                }, 400);
+              }
+              utils.map([ false, true ], function execute(prepare, mapNext) {
+                client.execute('SELECT * FROM system.local', null, { prepare: prepare, isIdempotent: true }, mapNext);
+              }, function (err, results) {
+                assert.ifError(err);
+                assert.strictEqual(results.length, 2);
+                var expectedHost;
+                if (policy instanceof ConstantSpeculativeExecutionPolicy) {
```

Comment:
It would be nice if there were a way to track whether or not there were multiple speculative executions executed on a request like a `speculativeExecutions` field which just simply had the number of executions.  This would offer a way to track whether or not speculative executions were happening.  For example, I see that you can't easily tell if a request was submitted to another host if we get a response for the first request first since it won't show up in `triedHosts` (from what I can tell so far).

Reply:
Hrmm maybe i'm wrong about this.  I had a configuration issue where not all queries being sent were idempotent (so some weren't speculatively retried).  It looks like triedHosts would represent ones where speculative executions have begun, so perhaps a separate identifier isn't needed, unless we wanted to be able to distinguish between retries and speculative executions.

Reply:
I made a patch for helping me count speculative execution ratio, not sure if worth incorporating the idea: https://gist.github.com/tolbertam/69434846c4fcf5bb04f7f12783c5a8f1

Reply:
The patch looks good to me, push it!

Reply:
Will do, thanks!